#Networkit vs Networkx 

The purpouse of this notebook is to compare the eficciency of each library to compute some centrality measures: Degree, Beteweeness, Closeness, Eigenvector.

Each time a measure is computed you can observe the time it takes and a list containing the first 15 results of the ranking in format (node_id, measure_value)

######  Import the libraries and load the graph to be used

In [1]:
import networkit as nk
import networkx as nx
import operator

In [2]:
name = "../data/realNetworks/Power grid/power.gml"

In [3]:
g = nk.readGraph(name, nk.Format.GML)
gx = nk.nk2nx(g)

-----------------------------
### Betweenness

#### Compute betweenness using networkx

In [4]:
%time nxBet = nx.betweenness_centrality(gx, normalized=True)
sorted(nxBet.items(), key=operator.itemgetter(1), reverse=True)[:15]

CPU times: user 3min 1s, sys: 337 ms, total: 3min 2s
Wall time: 3min 2s


[(4164, 0.2884156214793963),
 (2543, 0.281698123316245),
 (1243, 0.2796951897344535),
 (4219, 0.27764988777849753),
 (2528, 0.2673044750046778),
 (1267, 0.24827101665302148),
 (1308, 0.23651669104994388),
 (1244, 0.20523301004236533),
 (426, 0.20495396256691495),
 (2606, 0.20311626353593557),
 (2594, 0.19230583355335712),
 (2605, 0.1862957329429392),
 (69, 0.18412340121277324),
 (108, 0.17741564211788427),
 (1167, 0.17112033945509128)]

###### Compute betweenness using networkit

In [5]:
%time betweenness = nk.centrality.Betweenness(g, normalized=True).run()
nkBet = betweenness.ranking()
sorted(nkBet, key=operator.itemgetter(1), reverse=True)[:15]

CPU times: user 12.7 s, sys: 16.7 ms, total: 12.7 s
Wall time: 4.59 s


[(4164, 0.5768312429587938),
 (2543, 0.5633962466324896),
 (1243, 0.5593903794688999),
 (4219, 0.555299775556987),
 (2528, 0.5346089500093508),
 (1267, 0.49654203330605307),
 (1308, 0.4730333820999036),
 (1244, 0.41046602008472566),
 (426, 0.4099079251338286),
 (2606, 0.40623252707187385),
 (2594, 0.3846116671067201),
 (2605, 0.37259146588587067),
 (69, 0.36824680242554014),
 (108, 0.35483128423576576),
 (1167, 0.34224067891018484)]

    NK: total = 12.7s        NX: total = 3min 2s (183s)

Como podemos ver para calcular betweenness networkit es más eficiente que networkx en un orden de magnitud.

______________________________________________
### Eigenvector

###### Compute eigenvector using networkit

In [6]:
%time eigenvector = nk.centrality.EigenvectorCentrality(g).run()
nkEig = eigenvector.ranking()
sorted(nkEig, key=operator.itemgetter(1), reverse=True)[:15]

CPU times: user 1.46 s, sys: 6.67 ms, total: 1.47 s
Wall time: 546 ms


[(4381, 0.28664670703589046),
 (4345, 0.27353406103079825),
 (4336, 0.2475242881990209),
 (4332, 0.2430144751867278),
 (4352, 0.23741353099972934),
 (4384, 0.2225200847391362),
 (4402, 0.21524248212974462),
 (4347, 0.2140313072227409),
 (4395, 0.19680258964341416),
 (4373, 0.19374538952325013),
 (4398, 0.18059414552314085),
 (4361, 0.17973364704986852),
 (4401, 0.17793493769754218),
 (4413, 0.17320770724226722),
 (4408, 0.1660117901252523)]

###### Compute eigenvector using networkx

In [7]:
%time nxEig = nx.eigenvector_centrality(gx)
sorted(nxEig.items(), key=operator.itemgetter(1), reverse=True)[:15]

CPU times: user 857 ms, sys: 3.33 ms, total: 860 ms
Wall time: 865 ms


[(4381, 0.28654779326719626),
 (4345, 0.27441484972069885),
 (4336, 0.24815058029004833),
 (4332, 0.2429173678578642),
 (4352, 0.23638918982053217),
 (4384, 0.2214314447525528),
 (4402, 0.21480827911882583),
 (4347, 0.21410511011261296),
 (4395, 0.19772372883577932),
 (4373, 0.19468152174720804),
 (4361, 0.1805642233235685),
 (4398, 0.18033539417727554),
 (4401, 0.17708588734382957),
 (4413, 0.17247182475629233),
 (4408, 0.16532820596420036)]

    NK: total = 1.47s        NX: total = 860 ms

Networkx puede calcular betweenness aproximadamente 2 veces más rápido que networkit.

______________________________________________
### Degree

###### Degree centrality using networkx

In [8]:
%time nxDeg = nx.degree_centrality(gx)
sorted(nxDeg.items(), key=operator.itemgetter(1), reverse=True)[:15]

CPU times: user 3.33 ms, sys: 3.33 ms, total: 6.67 ms
Wall time: 7.2 ms


[(2553, 0.003846153846153846),
 (4458, 0.0036437246963562753),
 (831, 0.0028340080971659917),
 (3468, 0.0028340080971659917),
 (4345, 0.0028340080971659917),
 (2382, 0.002631578947368421),
 (2542, 0.002631578947368421),
 (2575, 0.002631578947368421),
 (2585, 0.002631578947368421),
 (3895, 0.002631578947368421),
 (1224, 0.00242914979757085),
 (2434, 0.00242914979757085),
 (2439, 0.00242914979757085),
 (2617, 0.00242914979757085),
 (2662, 0.00242914979757085)]

###### Degree centrality using networkit

In [9]:
%time degree = nk.centrality.DegreeCentrality(g).run()
nkDeg = degree.ranking()
sorted(nkDeg, key=operator.itemgetter(1), reverse=True)[:15]

CPU times: user 10 ms, sys: 0 ns, total: 10 ms
Wall time: 4.54 ms


[(2553, 19.0),
 (4458, 18.0),
 (831, 14.0),
 (4345, 14.0),
 (3468, 14.0),
 (2542, 13.0),
 (2382, 13.0),
 (2575, 13.0),
 (3895, 13.0),
 (2585, 13.0),
 (1224, 12.0),
 (2434, 12.0),
 (2439, 12.0),
 (2662, 12.0),
 (2617, 12.0)]

    NK: total = 10 ms        NX: total = 3.33 ms

Networkx puede calcular la centralidad de grado aproximadamente 3 veces más rápido que networkit.

______________________________________________
### Closeness

###### Closeness centrality using networkx

In [10]:
%time nxClos = nx.closeness_centrality(gx)
sorted(nxClos.items(), key=operator.itemgetter(1), reverse=True)[:15]

CPU times: user 43.4 s, sys: 3.33 ms, total: 43.4 s
Wall time: 43.4 s


[(1308, 0.08182330142114155),
 (2594, 0.0809437981320662),
 (2605, 0.08035786905246035),
 (1131, 0.07987194619153098),
 (2606, 0.0796260477111541),
 (1243, 0.07932684587468285),
 (1476, 0.07918570169111165),
 (2557, 0.07911468426194328),
 (2528, 0.07877280265339967),
 (2532, 0.07803121248499399),
 (2612, 0.07802751496580373),
 (2543, 0.07802628253727571),
 (4219, 0.07729377894605081),
 (1125, 0.07720680170042511),
 (2618, 0.07718388200553099)]

###### Closeness centrality using networkit

In [11]:
%time closeness = nk.centrality.Closeness(g).run()
nkClo = closeness.ranking()
sorted(nkClo, key=operator.itemgetter(1), reverse=True)[:15]

CPU times: user 6.59 s, sys: 0 ns, total: 6.59 s
Wall time: 2.37 s


[(1308, 1.6563421340312053e-05),
 (2594, 1.6385384237260363e-05),
 (2605, 1.6266775111834078e-05),
 (1131, 1.6168410160228943e-05),
 (2606, 1.6118633139909736e-05),
 (1243, 1.605806596653499e-05),
 (1476, 1.60294942694558e-05),
 (2557, 1.6015118271648438e-05),
 (2528, 1.594591146829953e-05),
 (2532, 1.579579200101093e-05),
 (2612, 1.5795043515344886e-05),
 (2543, 1.579479403588577e-05),
 (4219, 1.5646513956690448e-05),
 (1125, 1.5628907226806703e-05),
 (2618, 1.5624267612455665e-05)]

    NK: total = 6.59 s        NX: total = 43.4 s (3280 ms)

Networkit es aproximadamente 6 veces más rápido que networkx para calcular closeness.

##  Conclusion

Networkit es mas rápido calculando:
* Betweeness ( ~ 15x) 
* Closeness  ( ~ 6x)

Networkx es más eficiente para calcular:
* Degree ( ~ 3x)
* Eigenvector ( ~ 1.7x)

------ Falta formular porque la elección de networkx -----------